src/server/console/linux_virtual_terminal.h

```diff
@@ -75,22 +80,27 @@ class PosixProcessOperations
 class LinuxVirtualTerminal : public ConsoleServices
 {
 public:
-    LinuxVirtualTerminal(std::shared_ptr<VTFileOperations> const& fops,
-                         std::unique_ptr<PosixProcessOperations> pops,
-                         int vt_number,
-                         std::shared_ptr<graphics::DisplayReport> const& report);
+    LinuxVirtualTerminal(
+        std::shared_ptr<VTFileOperations> const& fops,
+        std::unique_ptr<PosixProcessOperations> pops,
+        int vt_number,
+        EmergencyCleanupRegistry& emergency_cleanup,
+        std::shared_ptr<graphics::DisplayReport> const& report);
     ~LinuxVirtualTerminal() noexcept(true);
 
     void register_switch_handlers(
         graphics::EventHandlerRegister& handlers,
         std::function<bool()> const& switch_away,
         std::function<bool()> const& switch_back) override;
     void restore() override;
-    std::future<std::unique_ptr<Device>> acquire_device(
+    std::future<std::unique_ptr<mir::Device>> acquire_device(
         int major, int minor,
-        std::unique_ptr<Device::Observer> observer) override;
+        std::unique_ptr<mir::Device::Observer> observer) override;
 
+    class Device;
 private:
+    std::shared_ptr<Mutex<std::vector<Device*>>> const active_devices;
```

Comment:
That makes my eyes bleed!

Reply:
Seems fairly sensible to me - it's a list of Device*'s that's safely mutatable from multiple threads, with shared ownership :grin: 

I've folded this up into a DeviceList class. Better?